integration-tests/src/main/java/org/apache/druid/testing/utils/QueryResultVerifier.java

```diff
@@ -19,13 +19,19 @@
 
 package org.apache.druid.testing.utils;
 
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.java.util.common.logger.Logger;
+
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 
 public class QueryResultVerifier
 {
-  public static boolean compareResults(
+  private static final Logger LOG = new Logger(QueryResultVerifier.class);
+
+  public static Optional<String> compareResults(
```

Comment:
Having an Optional of error message is better than just the boolean. But the best thing to do in such cases is to have a dedicated response object, which contains a success status and an error message. But since this is for tests only, we can do it later.

Reply:
That was my initial approach, however, I realised that if success is true, then the error message would be null and if  success is false then the error message would be some nonnull value, therefore having the success status seemed redundant to me. I can make the desired change while addressing the review comments.

Reply:
Yes, within the response object, you may choose to maintain a single String that represents the message. So while `getMessage()` would simply return that message, `isSuccess()` could do a null check on it.
I would still prefer having a separate `success` flag as you might later want to include a message even in cases of success.

Reply:
The approach taken in the planner test framework is to define a "test case" that has the query and expected results. Run that to get the actual results. Compare the two. The idea is to focus on the test and make the test code as simple as possible. Here:

```java
testCase = loadTestCase(...);
result = testCase.run(...);
```

Where `run()` would verify the results and would fail if the results don't match. (Results could be an expected error, or expected rows.)

Reply:
I would defer against the suggestion of `QueryResultVerifier.assertEqual(...);` because the helper is not going to do assertions based on the result. This is because the same method is utilized in `ITQueryRetryTestOnMissingSegments.java` which depending on a flag fails or drops the mismatch in the rows. Therefore one way would be to create a separate method that helps perform the Assertions. But since this is not required in several places, I haven't created a such method.

 Good point regarding the test case in the planner. That would certainly be a cleaner and an easier approach to add tests.